internal/server/reports.go

```diff
@@ -66,18 +75,175 @@ func (s *Server) putReport() http.HandlerFunc {
 		}
 		report.RealmID = &realmID
 
-		created, err := s.Store.UpsertReport(r.Context(), report)
+		// make sure team is present at match, and the event is visible to user
+		present, err := s.Store.IsTeamPresentAtMatch(r.Context(), report.EventKey, report.MatchKey, report.TeamKey, &realmID)
```

Comment:
This is another one of those annoying scenarios where checking whether the user has access to the resource and making changes to the resource should happen in a transaction or something to make sure their access stays the same while it's changed.